lib/srv/db/common/packettrace/tracing.go

```diff
@@ -0,0 +1,235 @@
+	for _, packet := range packets {
+		// assign indicator or skip packet.
+		indicator := ""
+		if packet.Direction == sender.direction {
+			indicator = "I"
+		}
+		if packet.Direction == receiver.direction {
+			indicator = "O"
+		}
+		if indicator == "" {
+			continue
+		}
+
+		// Write the timestamp and sender indicator
+		if _, err := fmt.Fprintf(buffer, "%s %s\n", indicator, packet.Timestamp.UTC().Format(time.RFC3339Nano)); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Write the packet data using hex.Dump and add a newline after each packet
+		hexData := hex.Dump(packet.Payload)
+		if _, err := buffer.Write([]byte(hexData + "\n")); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	filenameHex := filename + ".hex"
+
+	err := os.WriteFile(filenameHex, buffer.Bytes(), 0600)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer os.Remove(filenameHex)
+
+	// Invoke text2pcap to convert the hex dump to PCAP format
+	addrPair := fmt.Sprintf("%v,%v", sender.addr, receiver.addr)
+	out, err := t.runCommand(text2pcapBin,
+		// enable inbound/outbound markers
+		"-D",
+		// configure timestamp format
+		"-t", "%Y-%m-%dT%H:%M:%S.%fZ",
+		// ethernet capture encapsulation type
+		"-l", "1",
+		// specify IPv4 addr pair
+		"-4", addrPair,
+		// inbound/outbound ports; also used to hint at packet type; might be worth using "-P <dissector>" instead.
+		"-T", fmt.Sprintf("%v,%v", port, port),
+		filenameHex,
+		filename)
+	if err != nil {
+		return trace.Wrap(err, "error running text2pcap (output: %v)", out)
+	}
+
+	return nil
+}
+
+// SaveToPCAP saves the trace to a single merged pcap file.
+func (t *Trace) SaveToPCAP(baseFilename string, port int) error {
+	// Skip saving if filename is empty.
+	if baseFilename == "" {
+		return nil
+	}
+
+	t.mx.Lock() // Lock the mutex for the duration of the packet writing
+	defer t.mx.Unlock()
+
+	// we write to .pcap files:
+	// client <-> teleport
+	filename001 := baseFilename + ".001"
+	clientPart := participant{addr: fakeClientAddr, direction: ClientToTeleport}
+	teleportPart1 := participant{addr: fakeTeleportAddr, direction: TeleportToClient}
+	err := t.saveOneLinkToPCAP(filename001, port, clientPart, teleportPart1, t.packets)
+	if err != nil {
+		return trace.Wrap(err, "error saving to PCAP")
+	}
+	defer os.Remove(filename001)
+
+	// teleport <-> server
+	filename002 := baseFilename + ".002"
+	teleportPart2 := participant{addr: fakeTeleportAddr, direction: TeleportToServer}
+	serverPart := participant{addr: fakeServerAddr, direction: ServerToTeleport}
+	err = t.saveOneLinkToPCAP(filename002, port, teleportPart2, serverPart, t.packets)
+	if err != nil {
+		return trace.Wrap(err, "error saving to PCAP")
+	}
+	defer os.Remove(filename002)
+
+	// merge two files
+	out, err := t.runCommand(mergecapBin, "-w", baseFilename, filename001, filename002)
+	if err != nil {
+		return trace.Wrap(err, "error running mergecap (output: %v)", out)
+	}
+	return nil
+}
+
+func (t *Trace) SaveAsText(file string) error {
```

Comment:
Should we implement `io.WriterTo` instead for more flexibility and interoperability?

Reply:
I can add this other method and have `SaveAsText` call it.

Reply:
Added.